# Reproducible Workflows for Genomics Research

This repository contains file descriptions for common bioinformatics tools and workflows.
These workflows have been developed and used by the Biostatistics and Computational Laboratory (BCBLab).
Descriptions of tools and workflows were written using reproducible languages for reproducible research:
Common Workflow Language (CWL) and Workflow Description Language (WDL).
Newest versions are mainly written in WDL.

## Tools

Dockerfiles and CWL/WDL files for tools are in `tools` directory.
Each subdirectory for one tool.
Inside each subdirectory there are many directories representing tool version.
The version must be equal to tool version downloaded by Dockerfile.
Dockerfiles are tagged according to tool version.
Each version must have its own Dockerfile and CWL/WDL files.
CWL/WDL must use the right version of Docker image according to tool version.

| Tool          | Versions                   | Description                                                                                                       |
| ------------- | -------------------------- | ----------------------------------------------------------------------------------------------------------------- |
| admixture     | 1.3.0                      | Software tool for maximum likelihood estimation of individual ancestries from multilocus SNP genotype datasets    |
| bcl2fastq     | 1.8.4, 2.20.0              | Demultiplexes data and converts BCL files generated by Illumina sequencing systems to standard FASTQ file formats |
| Bioconductor  | 3.6, 3.7, 3.9              | Provides tools for the analysis and comprehension of high- throughput genomic data                                |
| Bismark       | 0.14.5, 0.19.0, 0.22.1     | Maps bisulfite treated sequencing reads to a genome of interest and perform methylation calls in a single step    |
| Bowtie        | 1.2.0, 1.2.2, 1.2.3        | An ultrafast, memory-efficient short read aligner                                                                 |
| Bowtie 2      | 2.3.5.1                    | An ultrafast and memory-efficient tool for aligning sequencing reads to long reference sequences                  |
| BWA           | 0.7.15, 0.7.17             | Burrows-Wheeler Aligner                                                                                           |
| Cutadapt      | 1.13, 1.16, 2.3            | Removes adapter sequences from sequencing reads                                                                   |
| FastQC        | 0.11.6, 0.11.7, 0.11.8     | A quality control tool for high throughput sequence data                                                          |
| fineSTRUCTURE | 2.1.1                      | Fast and powerful algorithm for identifying population structure using dense sequencing data                      |
| GATK          | 3.6-0                      | Genome Analysis Toolkit - Variant Discovery in High-Throughput Sequencing Data                                    |
| HTSeq-count   | 0.9.1, 0.11.1              | Counting reads in features                                                                                        |
| Reaper        | 13-274                     | Program for demultiplexing, trimming and filtering short read sequencing data                                     |
| Picard        | 2.6.0                      | A set of command line tools for manipulating high-throughput sequencing data                                      |
| Plink         | 1.07, 1.09                 | Whole genome association analysis toolset                                                                         |
| R             | 3.4.2, 3.5.0/2, 3.6.1      | Language and environment for statistical computing and graphics                                                   |
| Rqc           | 1.14.0, 1.16.2             | Quality Control Tool for High-Throughput Sequencing Data                                                          |
| Samtools      | 1.5, 1.9                   | Reading/writing/editing/indexing/viewing SAM/BAM/CRAM format                                                      |
| Shapeit2      | 900                        | Fast and accurate method for estimation of haplotypes (aka phasing) from genotype or sequencing data              |
| STAR          | 2.5.3a, 2.7.1a             | RNA-seq aligner                                                                                                   |
| featureCounts | 1.5.2, 1.6.2, 1.6.4        | Counting reads to genomic features such as genes, exons, promoters and genomic bins                               |
| TrimGalore!   | 0.4.4, 0.4.5, 0.5.0, 0.6.3 | A wrapper tool around Cutadapt and FastQC to consistently apply quality and adapter trimming to FastQ files       |
| VCFTools      | 0.1.16                     | Program package designed for working with VCF files                                                               |


## Workflows

Workflows import tool descriptions allowing reuse of these files.

| Workflow    | Versions                      | Description                                     |
| ----------- | ----------------------------- | ----------------------------------------------- |
| BAM to CRAM | 1.0.0, 1.1.0                  | Convert BAM files to CRAM format                |
| miRNA-seq   | 1.0, 1.1, 1.2.0               | Quantification of known microRNAs from NGS data |
| QA          | 1.0, 1.1                      | Quality assessment of NGS data                  |
| RNA-seq     | 1.0, 1.1, 1.2.0, 1.3.0, 2.0.0 | Quantification of known genes from NGS data     |
| WGBS        | 1.0, 1.1, 1.2.0               | Quantification of CpG methylation from NGS data |

## Build  Docker images and push to DockerHub

The command below builds and pushes Docker images to DockerHub.
Images are tagged with tool versions.

```bash
bash build_images.sh
```

## Generate inputs files

Generating input files make WOMTOOL to validate WDL workflow and tools files.

```bash
bash generate_inputs.sh
```
